src/isl/isl_range.rs

```diff
@@ -0,0 +1,850 @@
+            }
+            (TypedRangeBoundaryValue::Min, TypedRangeBoundaryValue::Number(v2)) => {
+                RangeImpl::range(RangeBoundaryValue::Min, v2)
+            }
+            (TypedRangeBoundaryValue::Number(v1), TypedRangeBoundaryValue::Max) => {
+                RangeImpl::range(v1, RangeBoundaryValue::Max)
+            }
+            (TypedRangeBoundaryValue::Number(Value(v1, _)), _) => {
+                invalid_schema_error("Range boundaries should have same types")
+            }
+            _ => unreachable!(
+                "Number ranges can not be constructed with non number range boundary types"
+            ),
+        }
+    }
+
+    /// Provides `RangeImpl<TimestampPrecision>` for given `TypedRangeBoundaryValue`
+    // this method requires a prior check for TypedRangeBoundaryValue::TimestampPrecision
+    pub(crate) fn timestamp_precision_range_from_typed_boundary_value(
+        start: TypedRangeBoundaryValue,
+        end: TypedRangeBoundaryValue,
+    ) -> IonSchemaResult<RangeImpl<TimestampPrecision>> {
+        match (start, end) {
+            (TypedRangeBoundaryValue::TimestampPrecision(v1), TypedRangeBoundaryValue::TimestampPrecision(v2)) => {
+                RangeImpl::range(v1, v2)
+            }
+            (TypedRangeBoundaryValue::Min, TypedRangeBoundaryValue::TimestampPrecision(v2)) => {
+                RangeImpl::range(RangeBoundaryValue::Min, v2)
+            }
+            (TypedRangeBoundaryValue::TimestampPrecision(v1), TypedRangeBoundaryValue::Max) => {
+                RangeImpl::range(v1, RangeBoundaryValue::Max)
+            }
+            (TypedRangeBoundaryValue::TimestampPrecision(Value(v1, _)), _) => {
+                invalid_schema_error("Range boundaries should have same types")
+            }
+            _ => unreachable!(
+                "TimestampPrecision ranges can not be constructed with non timestamp precision range boundary types"
+            ),
+        }
+    }
+
+    /// Provides `RangeImpl<Decimal>` for given `TypedRangeBoundaryValue`
+    // this method requires a prior check for TypedRangeBoundaryValue::Decimal
+    pub(crate) fn decimal_range_from_typed_boundary_value(
+        start: TypedRangeBoundaryValue,
+        end: TypedRangeBoundaryValue,
+    ) -> IonSchemaResult<RangeImpl<Decimal>> {
+        match (start, end) {
+            (TypedRangeBoundaryValue::Decimal(v1), TypedRangeBoundaryValue::Decimal(v2)) => {
+                RangeImpl::range(v1, v2)
+            }
+            (TypedRangeBoundaryValue::Min, TypedRangeBoundaryValue::Decimal(v2)) => {
+                RangeImpl::range(RangeBoundaryValue::Min, v2)
+            }
+            (TypedRangeBoundaryValue::Decimal(v1), TypedRangeBoundaryValue::Max) => {
+                RangeImpl::range(v1, RangeBoundaryValue::Max)
+            }
+            (TypedRangeBoundaryValue::Decimal(Value(v1, _)), _) => {
+                invalid_schema_error("Range boundaries should have same types")
```

Comment:
It feels like we should be able to eliminate some of the repetitive boilerplate in these `match` statements.